controllers/etcd_controller.go

```diff
@@ -159,30 +197,501 @@ func (r *EtcdReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	logger.Infof("Reconciling etcd services for etcd:%s in namespace:%s", etcd.Name, etcd.Namespace)
+
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: etcd.Spec.Labels,
+	})
+	if err != nil {
+		logger.Error(err, "Error converting etcd selector to selector")
+		return nil, err
+	}
+
+	// list all services to include the services that don't match the etcd`s selector
+	// anymore but has the stale controller ref.
+	services := &corev1.ServiceList{}
+	err = r.List(context.TODO(), services, client.InNamespace(etcd.Namespace))
+	if err != nil {
+		logger.Error(err, "Error listing statefulsets")
+		return nil, err
+	}
+
+	// NOTE: filteredStatefulSets are pointing to deepcopies of the cache, but this could change in the future.
+	// Ref: https://github.com/kubernetes-sigs/controller-runtime/blob/release-0.2/pkg/cache/internal/cache_reader.go#L74
+	// if you need to modify them, you need to copy it first.
+	filteredServices, err := r.claimServices(etcd, selector, services)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(filteredServices) > 0 {
+		// TODO: Sync spec and delete OR First delete and then sync spec?
+
+		// Keep only 1 Service. Delete the rest
+		for i := 1; i < len(filteredServices); i++ {
+			ss := filteredServices[i]
+			if err := r.Delete(context.TODO(), ss); err != nil {
+				logger.Error(err, "Error in deleting duplicate StatefulSet")
+				continue
+			}
+		}
+
+		// Return the updated Service
+		service := &corev1.Service{}
+		err = r.Get(context.TODO(), types.NamespacedName{Name: filteredServices[0].Name, Namespace: filteredServices[0].Namespace}, service)
+		if err != nil {
+			return nil, err
+		}
+		// Statefulset is claimed by for this etcd. Just sync the specs
+		if service, err = r.syncServiceSpec(service, etcd); err != nil {
+			return nil, err
+		}
+		return service, err
+	}
+
+	// Required Service doesn't exist. Create new
+
+	ss, err := r.getServiceFromEtcd(etcd)
+	if err != nil {
+		return nil, err
+	}
+
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		return r.Create(context.TODO(), ss)
+	})
+
+	// Ignore the precondition violated error, this machine is already updated
+	// with the desired label.
+	if err == errorsutil.ErrPreconditionViolated {
+		logger.Infof("Service %s precondition doesn't hold, skip updating it.", ss.Name)
+		err = nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := controllerutil.SetControllerReference(etcd, ss, r.Scheme); err != nil {
+		return nil, err
+	}
+
+	return ss.DeepCopy(), err
+}
+
+func (r *EtcdReconciler) syncServiceSpec(ss *corev1.Service, etcd *druidv1alpha1.Etcd) (*corev1.Service, error) {
+	decoded, err := r.getServiceFromEtcd(etcd)
+	if err != nil {
+		return nil, err
+	}
+
+	if reflect.DeepEqual(ss.Spec, decoded.Spec) {
+		return ss, nil
+	}
+	ssCopy := ss.DeepCopy()
+	decoded.Spec.DeepCopyInto(&ssCopy.Spec)
+	// Copy ClusterIP as the field is immutable
+	ssCopy.Spec.ClusterIP = ss.Spec.ClusterIP
```

Comment:
I'll error out if the type is changed. In case of ClusterIP, we need to sync the field as it is not in the service chart template. As for type: ClusterIP, we are explicitly setting it in the template. Ill make a note there to refrain from changing the type casually. WDYT?

Reply:
👍 Ok. We will deal with that problem when it is required.